hud/notifications.lua

```diff
@@ -0,0 +1,156 @@
+        local background = ar.quad(glasses, {x, y+1}, {x, y+10}, {x, y+10}, {x, y+1}, data.borderColor, 0.4)
+        local text = ar.text(glasses, string, {-width, y+2}, color or data.primaryColor)
+        local stepsTaken = 0
+        local totalSteps = width / stepSize
+        local direction = 1
+        local function advance()
+            stepsTaken = stepsTaken + direction
+            top.setVertex(3, x + stepSize * stepsTaken, y+1)
+            top.setVertex(4, x + stepSize * stepsTaken - 1, y)
+            bottom.setVertex(3, x + stepSize * stepsTaken/1.5 - 1, y+11)
+            bottom.setVertex(4, x + stepSize * stepsTaken/1.5, y+10)
+            background.setVertex(3, x + stepSize * stepsTaken - 8 - 1, y+10)
+            background.setVertex(4, x + stepSize * stepsTaken, y+1)
+            text.setPosition(math.min(x+1, x + -width + stepSize * stepsTaken + 1), y+2)
+            if direction == -1 and stepsTaken == 0 then
+                glasses.removeObject(top.getID())
+                glasses.removeObject(bottom.getID())
+                glasses.removeObject(background.getID())
+                glasses.removeObject(text.getID())
+            end
+        end
+        local function retract()
+            direction = -1
+            event.timer(0.05, advance, totalSteps)
+            data.notifications[yModifier] = false
+        end
+        event.timer(0.05, advance, totalSteps)
+        if timeout then event.timer(timeout, retract) else return retract end
+    else
+        table.insert(data.queue, {string, timeout, color})
+        return false
+    end
+end
+
+function notifications.addNotification(text, timeout, color)
+    local retracts = {}
+    for i = 1, #hudObjects do
+        local hudObject = hudObjects[i]
+        local retractFunc = notification(hudObject, text, timeout, color)
+        if not timeout then table.insert(retracts, retractFunc) end
+    end
+    local function retractAll()
+        --Remove notifications from queue if they were never shown
+        for i = 1, #retracts do
+            if retracts[i] == false then
+                for j = 1, #hudObjects do
+                    for k = 1, #hudObjects[j].queue do
+                        if hudObjects[j].queue[k][1] == text then
+                            table.remove(hudObjects[j].queue, k)
+                            break
+                        end
+                    end
+                end
+            else
+                retracts[i]()
+            end
+        end
+    end
+    if not timeout then return retractAll end
+end
+local function processQueue(hudObject)
+    local queue = hudObject.queue[1]
+    for i = 1, #hudObject.notifications do
+        if not hudObject.notifications[i] then
+            table.remove(hudObject.queue, 1)
+            notification(hudObject, queue[1], queue[2], queue[3])
+            break
+        end
+    end
+end
+
+function notifications.widget(glasses)
+    if #hudObjects < #glasses then
```

Comment:
Y?

Reply:
Copied code from rest of the HUD modules.

Basically just checks if the module has been initialized yet.